processor/src/decoder/trace.rs

```diff
@@ -0,0 +1,374 @@
+        self.hasher_trace[6].push(ZERO);
+        self.hasher_trace[7].push(ZERO);
+
+        let last_group_count = self.last_group_count();
+        debug_assert!(last_group_count == ZERO, "group count not zero");
+        self.group_count_trace.push(last_group_count);
+
+        self.op_idx_trace.push(ZERO);
+    }
+
+    // TRACE GENERATION
+    // --------------------------------------------------------------------------------------------
+
+    /// Returns vector of columns of this execution trace.
+    ///
+    /// Each column in the trace is extended to the specified trace length. The extension is done
+    /// by inserting ZEROs into the unfilled rows of most columns. The only exceptions are:
+    /// - The op_bits columns, where the unfilled rows are filled with the opcode of the HALT
+    ///   operation.
+    /// - The first 4 columns of the hasher state, where the unfilled rows are filled with the
+    ///   values from the last filled row. This is done so that the hash of the program is
+    ///   propagated to the last row.
+    pub fn into_vec(mut self, trace_len: usize, num_rand_rows: usize) -> Vec<Vec<Felt>> {
+        let own_len = self.trace_len();
+        // make sure that only the duplicate rows will be overwritten with random values
+        assert!(
+            own_len + num_rand_rows <= trace_len,
+            "target trace length too small"
+        );
+
+        let mut trace = Vec::new();
+
+        // put ZEROs into the unfilled rows of block address column
+        self.addr_trace.resize(trace_len, ZERO);
+        trace.push(self.addr_trace);
+
+        // insert HALT opcode into the unfilled rows of ob_bits columns
+        let halt_opcode = Operation::Halt.op_code().expect("missing opcode");
+        for (i, mut column) in self.op_bits_trace.into_iter().enumerate() {
+            debug_assert_eq!(own_len, column.len());
+            let value = Felt::from((halt_opcode >> i) & 1);
+            column.resize(trace_len, value);
+            trace.push(column);
+        }
+
+        // put ZEROs into the unfilled rows of in_span column
+        debug_assert_eq!(own_len, self.in_span_trace.len());
+        self.in_span_trace.resize(trace_len, ZERO);
+        trace.push(self.in_span_trace);
+
+        // for unfilled rows of hasher state columns, copy over values from the last row for the
+        // first 4 columns, and pad the other 4 columns with ZEROs
+        for (i, mut column) in self.hasher_trace.into_iter().enumerate() {
+            debug_assert_eq!(own_len, column.len());
+            if i < 4 {
+                // TODO: ideally, we should use `expect()` instead of `unwrap_or()` but aux_table
+                // unit tests fail if we do that as they don't expect build the decoder portion of
+                // the trace. in the future we should refactor aux_table unit tests.
+                let last_value = *column.last().unwrap_or(&ZERO);
+                column.resize(trace_len, last_value);
```

Comment:
This doesn't sound like it should be too hard to fix. Is it worth making an issue after this is merged?

Reply:
Yes - but it is not trivial either. So, yes - let's create an issue for this.